// three-way-merge/dist/modules/src/outcomes.js
export class Outcome {
    isResolved() {
        return !this.hasConflicts;
    }
    isConflicted() {
        return this.hasConflicts;
    }
}
export class Conflicted extends Outcome {
    constructor(left, base, right) {
        super();
        this.left = left;
        this.base = base;
        this.right = right;
        this.hasConflicts = true;
    }
    // Special constructor because left/base/right positional params
    // are confusing
    static create(opts) {
        return new Conflicted(opts.left, opts.base, opts.right);
    }
    apply(fun) {
        return Conflicted.create({
            left: fun(this.left),
            base: fun(this.base),
            right: fun(this.right)
        });
    }
}
export class Resolved extends Outcome {
    constructor(result) {
        super();
        this.hasConflicts = false;
        this.result = result;
    }
    combine(other) {
        this.result = this.result.concat(other.result);
    }
    apply(fun) {
        return new Resolved(fun(this.result));
    }
}
