// three-way-merge/dist/commonjs/src/diff3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const heckel_diff_1 = require("./heckel-diff");
class Diff2Command {
    constructor(code, baseLo, baseHi, sideLo, sideHi) {
        this.code = code;
        this.baseLo = baseLo;
        this.baseHi = baseHi;
        this.sideLo = sideLo;
        this.sideHi = sideHi;
    }
    static fromChangeRange(changeRange) {
        return new Diff2Command(changeRange.action, changeRange.leftLo, changeRange.leftHi, changeRange.rightLo, changeRange.rightHi);
    }
}
exports.Diff2Command = Diff2Command;
class Diff3 {
    constructor(left, base, right) {
        this.left = left;
        this.base = base;
        this.right = right;
    }
    static executeDiff(left, base, right) {
        return new Diff3(left, base, right).getDifferences();
    }
    getDifferences() {
        const leftDiff = heckel_diff_1.default.diff(this.base, this.left).map((d) => {
            return Diff2Command.fromChangeRange(d);
        });
        const rightDiff = heckel_diff_1.default.diff(this.base, this.right).map((d) => {
            return Diff2Command.fromChangeRange(d);
        });
        return this.collapseDifferences(new DiffDoubleQueue(leftDiff, rightDiff));
    }
    collapseDifferences(diffsQueue, differences = []) {
        if (diffsQueue.isFinished()) {
            return differences;
        }
        else {
            const resultQueue = new DiffDoubleQueue();
            const initSide = diffsQueue.chooseSide();
            const topDiff = diffsQueue.dequeue();
            resultQueue.enqueue(initSide, topDiff);
            diffsQueue.switchSides();
            this.buildResultQueue(diffsQueue, topDiff.baseHi, resultQueue);
            differences.push(this.determineDifference(resultQueue, initSide, diffsQueue.switchSides()));
            return this.collapseDifferences(diffsQueue, differences);
        }
    }
    buildResultQueue(diffsQueue, prevBaseHi, resultQueue) {
        if (this.queueIsFinished(diffsQueue.peek(), prevBaseHi)) {
            return resultQueue;
        }
        else {
            const topDiff = diffsQueue.dequeue();
            resultQueue.enqueue(diffsQueue.currentSide, topDiff);
            if (prevBaseHi < topDiff.baseHi) {
                diffsQueue.switchSides();
                return this.buildResultQueue(diffsQueue, topDiff.baseHi, resultQueue);
            }
            else {
                return this.buildResultQueue(diffsQueue, prevBaseHi, resultQueue);
            }
        }
    }
    queueIsFinished(queue, prevBaseHi) {
        return queue.length === 0 || queue[0].baseLo > prevBaseHi + 1;
    }
    determineDifference(diffDiffsQueue, initSide, finalSide) {
        const baseLo = diffDiffsQueue.get(initSide)[0].baseLo;
        const finalQueue = diffDiffsQueue.get(finalSide);
        const baseHi = finalQueue[finalQueue.length - 1].baseHi;
        const [leftLo, leftHi] = this.diffableEndpoints(diffDiffsQueue.get(Side.left), baseLo, baseHi);
        const [rightLo, rightHi] = this.diffableEndpoints(diffDiffsQueue.get(Side.right), baseLo, baseHi);
        const leftSubset = this.left.slice(leftLo - 1, leftHi);
        const rightSubset = this.right.slice(rightLo - 1, rightHi);
        const changeType = this.decideAction(diffDiffsQueue, leftSubset, rightSubset);
        return new Difference(changeType, leftLo, leftHi, rightLo, rightHi, baseLo, baseHi);
    }
    diffableEndpoints(commands, baseLo, baseHi) {
        if (commands.length) {
            const firstCommand = commands[0];
            const lastCommand = commands[commands.length - 1];
            const lo = firstCommand.sideLo - firstCommand.baseLo + baseLo;
            const hi = lastCommand.sideHi - lastCommand.baseHi + baseHi;
            return [lo, hi];
        }
        else {
            return [baseLo, baseHi];
        }
    }
    decideAction(diffDiffsQueue, leftSubset, rightSubset) {
        if (diffDiffsQueue.isEmpty(Side.left)) {
            return ChangeType.chooseRight;
        }
        else if (diffDiffsQueue.isEmpty(Side.right)) {
            return ChangeType.chooseLeft;
        }
        else {
            // leftSubset deepEquals rightSubset
            if (!leftSubset.every((x, i) => rightSubset[i] === x)) {
                return ChangeType.possibleConflict;
            }
            else {
                return ChangeType.noConflictFound;
            }
        }
    }
}
exports.default = Diff3;
class Difference {
    constructor(changeType, leftLo, leftHi, rightLo, rightHi, baseLo, baseHi) {
        this.changeType = changeType;
        this.leftLo = leftLo;
        this.leftHi = leftHi;
        this.rightLo = rightLo;
        this.rightHi = rightHi;
        this.baseLo = baseLo;
        this.baseHi = baseHi;
    }
}
exports.Difference = Difference;
var ChangeType;
(function (ChangeType) {
    ChangeType["chooseRight"] = "choose_right";
    ChangeType["chooseLeft"] = "choose_left";
    ChangeType["possibleConflict"] = "possible_conflict";
    ChangeType["noConflictFound"] = "no_conflict_found";
})(ChangeType = exports.ChangeType || (exports.ChangeType = {}));
var Side;
(function (Side) {
    Side["left"] = "left";
    Side["right"] = "right";
})(Side = exports.Side || (exports.Side = {}));
class DiffDoubleQueue {
    constructor(left = [], right = []) {
        this.diffs = { left: left, right: right };
    }
    dequeue(side = this.currentSide) {
        return this.diffs[side].shift();
    }
    peek(side = this.currentSide) {
        return this.diffs[side];
    }
    isFinished() {
        return this.isEmpty(Side.left) && this.isEmpty(Side.right);
    }
    enqueue(side = this.currentSide, val) {
        return this.diffs[side].push(val);
    }
    get(side = this.currentSide) {
        return this.diffs[side];
    }
    isEmpty(side = this.currentSide) {
        return this.diffs[side].length === 0;
    }
    switchSides(side = this.currentSide) {
        return this.currentSide = (side === Side.left) ? Side.right : Side.left;
    }
    chooseSide() {
        if (this.isEmpty(Side.left)) {
            this.currentSide = Side.right;
        }
        else if (this.isEmpty(Side.right)) {
            this.currentSide = Side.left;
        }
        else {
            this.currentSide = (this.get(Side.left)[0].baseLo <= this.get(Side.right)[0].baseLo ? Side.left : Side.right);
        }
        return this.currentSide;
    }
}
exports.DiffDoubleQueue = DiffDoubleQueue;
